src/packages/utils/src/things/json-rpc/json-rpc-base-types.ts

```diff
@@ -1,116 +1,62 @@
-import { bigIntToBuffer } from "../../utils";
-import { uintToBuffer } from "../../utils";
-
-import { BUFFER_EMPTY } from "../../utils/constants";
-
-export const strCache = new WeakMap();
-export const bufCache = new WeakMap();
-export const toStrings = new WeakMap();
-export const toBuffers = new WeakMap();
-
+import {
+  JsonRpcInputArg,
+  parseAndValidateStringInput,
+  parseAndValidateBigIntInput,
+  parseAndValidateNumberInput
+} from "./input-parsers";
 const inspect = Symbol.for("nodejs.util.inspect.custom");
 
-export class BaseJsonRpcType<
-  T extends number | bigint | string | Buffer =
-    | number
-    | bigint
-    | string
-    | Buffer
-> {
-  public [Symbol.toStringTag]: string;
+export class BaseJsonRpcType {
+  protected bufferValue: Buffer | null;
 
-  protected value: T;
   // used to make console.log debugging a little easier
-  private [inspect](_depth: number, _options: any): T {
-    return this.value;
+  private [inspect](_depth: number, _options: any): string {
+    return `[${this.constructor.name}] ${this.toString()}`;
   }
-  constructor(value: T) {
-    const self = this;
-    if (Buffer.isBuffer(value)) {
-      toStrings.set(this, () => value.toString("hex"));
-      bufCache.set(this, value);
-      self[Symbol.toStringTag] = "Buffer";
+
+  constructor(value: JsonRpcInputArg) {
+    if (value == null) {
+      this.bufferValue = null;
+    } else if (Buffer.isBuffer(value)) {
+      // empty buffer should be treated as null
+      this.bufferValue = value.length === 0 ? null: value;
```

Comment:
Could this be consolidated into something like

``` javascript
if (value === null || (Buffer.isBuffer(value) && value.length === 0) {
  // codepath for nulls
  this.bufferValue = null;
} else {
 // codepath for non-null
 this.bufferValue = value;
}

```

Reply:
It could be - but it makes the branching more complicated, which I'm not a huge fan of (also we'd need to either check Buffer.isBuffer twice, or store that value).

Open to other POV though!